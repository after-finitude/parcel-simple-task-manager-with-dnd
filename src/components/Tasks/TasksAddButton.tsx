import React, { useCallback } from "react";
import { useDispatch } from "react-redux";
import { styled } from "@material-ui/core/styles";
import AddIcon from "@material-ui/icons/Add";
import Button from "@material-ui/core/Button";

import { openDialog } from "store/dialog/actions";
import { DialogTypes } from "store/dialog/types";

export const TasksAddButton: React.FC = () => {
  const dispatch = useDispatch();

  const handleAddButton = useCallback(
    () => dispatch(openDialog(DialogTypes.CREATE)),
    [dispatch],
  );

  return <AddButton onClick={handleAddButton}>Add task</AddButton>;
};

const AddButton = styled(props => (
  <Button startIcon={<AddIcon />} {...props} />
))({
  width: "100%",
});
