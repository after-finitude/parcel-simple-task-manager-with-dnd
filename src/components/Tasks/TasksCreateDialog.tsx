import React, { useCallback, memo } from "react";
import { useDispatch } from "react-redux";
import { useForm } from "react-hook-form";
import Button from "@material-ui/core/Button";
import TextField from "@material-ui/core/TextField";
import Dialog from "@material-ui/core/Dialog";
import DialogActions from "@material-ui/core/DialogActions";
import DialogContent from "@material-ui/core/DialogContent";
import DialogTitle from "@material-ui/core/DialogTitle";

import { closeDialog } from "store/dialog/actions";
import { DialogTypes } from "store/dialog/types";
import { thunkAddTask } from "store/tasks/thunks";

type FormData = {
  title: string;
  description: string;
};

interface Props {
  status?: boolean;
}

export const TasksCreateDialog: React.FC<Props> = memo(({ status = false }) => {
  const dispatch = useDispatch();
  const { register, handleSubmit } = useForm<FormData>();

  // eslint-disable-next-line react-hooks/exhaustive-deps
  const onSubmit = useCallback(
    handleSubmit(({ title, description }) => {
      dispatch(thunkAddTask(title, description));
    }),
    [dispatch, thunkAddTask],
  );
  const handleClose = useCallback(
    () => dispatch(closeDialog(DialogTypes.CREATE)),
    [dispatch],
  );

  return (
    <Dialog
      onClose={handleClose}
      open={status}
      aria-labelledby="form-dialog-title"
    >
      <DialogTitle id="form-dialog-title">Create task</DialogTitle>
      <form onSubmit={onSubmit}>
        <DialogContent>
          <TextField
            required
            autoFocus
            margin="dense"
            name="title"
            label="Title"
            variant="outlined"
            fullWidth
            inputRef={register}
          />
          <TextField
            margin="dense"
            name="description"
            label="Description"
            variant="outlined"
            fullWidth
            inputRef={register}
          />
        </DialogContent>
        <DialogActions>
          <Button onClick={handleClose} color="primary">
            Cancel
          </Button>
          <Button type="submit" color="primary">
            Add task
          </Button>
        </DialogActions>
      </form>
    </Dialog>
  );
});
