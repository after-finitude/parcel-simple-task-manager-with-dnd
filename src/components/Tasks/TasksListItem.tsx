import React, { memo, useCallback } from "react";
import { useDispatch } from "react-redux";
import styled from "styled-components";
import IconButton from "@material-ui/core/IconButton";
import CloseIcon from "@material-ui/icons/Close";
import EditIcon from "@material-ui/icons/Edit";
import RestoreIcon from "@material-ui/icons/Restore";
import ListItem from "@material-ui/core/ListItem";
import ListItemText from "@material-ui/core/ListItemText";
import ListItemSecondaryAction from "@material-ui/core/ListItemSecondaryAction";

import { openDialog } from "store/dialog/actions";
import { DialogTypes } from "store/dialog/types";
import { closeTask, restoreTask } from "store/tasks/actions";

interface ComponentProps {
  id: string;
  title: string;
  description?: string;
  closed: boolean;
}

export const TasksListItem: React.FC<ComponentProps> = memo(
  ({ id, title, description = "", closed }) => {
    const dispatch = useDispatch();

    const handleEdit = useCallback(
      () => dispatch(openDialog(DialogTypes.UPDATE, id)),
      [dispatch, id],
    );
    const handleClose = useCallback(() => dispatch(closeTask(id)), [
      dispatch,
      id,
    ]);
    const handleRestore = useCallback(() => dispatch(restoreTask(id)), [
      dispatch,
      id,
    ]);

    return (
      <ListItemWrapper closed={closed}>
        <ListItem>
          <ListItemText primary={title} secondary={description} />
          <ListItemSecondaryAction>
            <IconButton
              disabled={closed}
              onClick={handleEdit}
              edge="end"
              aria-label="edit"
            >
              <EditIcon />
            </IconButton>
            {closed ? (
              <IconButton
                onClick={handleRestore}
                edge="end"
                aria-label="restore"
              >
                <RestoreIcon />
              </IconButton>
            ) : (
              <IconButton onClick={handleClose} edge="end" aria-label="close">
                <CloseIcon />
              </IconButton>
            )}
          </ListItemSecondaryAction>
        </ListItem>
      </ListItemWrapper>
    );
  },
);

const ListItemWrapper = styled.div<{ closed: boolean }>`
  border: 1px grey solid;
  margin-bottom: 5px;
  opacity: ${(props: { closed: boolean }) => (props.closed ? 0.5 : 1)};
`;
