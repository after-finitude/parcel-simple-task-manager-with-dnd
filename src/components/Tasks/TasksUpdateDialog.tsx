import React, { useCallback, memo } from "react";
import { useDispatch, useSelector } from "react-redux";
import { useForm } from "react-hook-form";

import Button from "@material-ui/core/Button";
import TextField from "@material-ui/core/TextField";
import Dialog from "@material-ui/core/Dialog";
import DialogActions from "@material-ui/core/DialogActions";
import DialogContent from "@material-ui/core/DialogContent";
import DialogTitle from "@material-ui/core/DialogTitle";

import { RootState } from "store/rootReducer";
import { DialogTypes } from "store/dialog/types";
import { closeDialog } from "store/dialog/actions";
import { selectTaskById } from "store/dialog/selectors";
import { thunkUpdateTask } from "store/tasks/thunks";
import { Task } from "store/tasks/types";

type FormData = {
  title: string;
  description: string;
};

interface Props {
  status: boolean;
  currentTaskId: string;
}

export const TasksUpdateDialog: React.FC<Props> = memo(
  ({ status, currentTaskId }) => {
    const dispatch = useDispatch();
    const { register, handleSubmit } = useForm<FormData>();
    const currentTask = useSelector<RootState, Task | undefined>(
      selectTaskById(currentTaskId)
    );

    const onSubmit = useCallback(
      handleSubmit(({ title, description }) => {
        dispatch(thunkUpdateTask(title, description));
      }),
      [dispatch, thunkUpdateTask]
    );
    const handleClose = useCallback(
      () => dispatch(closeDialog(DialogTypes.UPDATE)),
      [dispatch]
    );

    return (
      <Dialog
        onClose={handleClose}
        open={status}
        aria-labelledby="form-dialog-title"
      >
        <DialogTitle id="form-dialog-title">Update task</DialogTitle>
        <form onSubmit={onSubmit}>
          <DialogContent>
            <TextField
              defaultValue={currentTask?.title ?? ""}
              autoFocus
              margin="dense"
              name="title"
              label="Title"
              variant="outlined"
              fullWidth
              inputRef={register}
            />
            <TextField
              defaultValue={currentTask?.description ?? ""}
              margin="dense"
              name="description"
              label="Description"
              variant="outlined"
              fullWidth
              inputRef={register}
            />
          </DialogContent>
          <DialogActions>
            <Button onClick={handleClose} color="primary">
              Cancel
            </Button>
            <Button type="submit" color="primary">
              Update task
            </Button>
          </DialogActions>
        </form>
      </Dialog>
    );
  }
);
