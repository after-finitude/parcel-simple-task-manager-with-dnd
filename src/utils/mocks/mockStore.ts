import configureStore from "redux-mock-store";

const store = configureStore();

export const mockStore = store({
  tasks: [
    { id: "test1", title: "Test1", description: "test1", closed: false },
    { id: "test2", title: "Test2", description: "test2", closed: false },
  ],
  dialog: { type: "", status: false, currentTaskId: "" },
});
