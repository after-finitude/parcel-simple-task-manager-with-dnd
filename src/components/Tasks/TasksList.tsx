import React, { useState, useEffect, memo, useCallback } from "react";
import { useSelector, useDispatch, shallowEqual } from "react-redux";
import { ReactSortable } from "react-sortablejs";
import List from "@material-ui/core/List";

import { RootState } from "store/rootReducer";
import { TasksState, Task } from "store/tasks/types";
import { updateState } from "store/tasks/actions";

import { TasksListItem } from "components/Tasks/TasksListItem";

export const TasksList: React.FC = memo(() => {
  const dispatch = useDispatch();
  const storeTasks = useSelector<RootState, TasksState>(
    (state) => state.tasks,
    shallowEqual
  );
  const [tasks, setTasks] = useState<TasksState>(storeTasks);

  useEffect(() => {
    setTasks(storeTasks);
  }, [storeTasks]);

  const onEnd = useCallback(() => dispatch(updateState(tasks)), [
    dispatch,
    tasks,
  ]);

  return (
    <List>
      <ReactSortable onEnd={onEnd} list={tasks} setList={setTasks}>
        {tasks.map((task: Task) => (
          <TasksListItem
            key={task?.id}
            id={task?.id}
            title={task?.title}
            description={task?.description}
            closed={task?.closed}
          />
        ))}
      </ReactSortable>
    </List>
  );
});
