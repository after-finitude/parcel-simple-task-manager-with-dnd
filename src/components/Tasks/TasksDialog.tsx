import React from "react";
import { useSelector, shallowEqual } from "react-redux";

import { TasksCreateDialog } from "components/Tasks/TasksCreateDialog";
import { TasksUpdateDialog } from "components/Tasks/TasksUpdateDialog";

import { RootState } from "store/rootReducer";
import { DialogState, DialogTypes } from "store/dialog/types";

export const TasksDialog: React.FC = () => {
  const dialog = useSelector<RootState, DialogState>(
    state => state.dialog,
    shallowEqual,
  );

  if (dialog.type === DialogTypes.CREATE) {
    return <TasksCreateDialog status={dialog.status} />;
  }

  return (
    <TasksUpdateDialog
      status={dialog.status}
      currentTaskId={dialog.currentTaskId}
    />
  );
};
