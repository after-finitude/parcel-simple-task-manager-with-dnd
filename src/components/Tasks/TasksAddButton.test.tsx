import React from "react";
import { Provider } from "react-redux";
import { mount } from "enzyme";

import { mockStore } from "utils/mocks/mockStore";
import { TasksAddButton } from "components/Tasks/TasksAddButton";

describe("TasksAddButton", () => {
  it("should render correctly", () => {
    const component = mount(
      <Provider store={mockStore}>
        <TasksAddButton />
      </Provider>,
    );

    expect(component).toMatchSnapshot();
  });
});
