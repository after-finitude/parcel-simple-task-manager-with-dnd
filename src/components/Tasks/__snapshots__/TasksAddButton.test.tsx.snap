// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TasksAddButton should render correctly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <TasksAddButton>
    <Styled(Component)
      onClick={[Function]}
    >
      <Component
        className="Component-root-1"
        onClick={[Function]}
      >
        <WithStyles(ForwardRef(Button))
          className="Component-root-1"
          onClick={[Function]}
          startIcon={<Memo(AddIcon) />}
        >
          <ForwardRef(Button)
            className="Component-root-1"
            classes={
              Object {
                "colorInherit": "MuiButton-colorInherit",
                "contained": "MuiButton-contained",
                "containedPrimary": "MuiButton-containedPrimary",
                "containedSecondary": "MuiButton-containedSecondary",
                "containedSizeLarge": "MuiButton-containedSizeLarge",
                "containedSizeSmall": "MuiButton-containedSizeSmall",
                "disableElevation": "MuiButton-disableElevation",
                "disabled": "Mui-disabled",
                "endIcon": "MuiButton-endIcon",
                "focusVisible": "Mui-focusVisible",
                "fullWidth": "MuiButton-fullWidth",
                "iconSizeLarge": "MuiButton-iconSizeLarge",
                "iconSizeMedium": "MuiButton-iconSizeMedium",
                "iconSizeSmall": "MuiButton-iconSizeSmall",
                "label": "MuiButton-label",
                "outlined": "MuiButton-outlined",
                "outlinedPrimary": "MuiButton-outlinedPrimary",
                "outlinedSecondary": "MuiButton-outlinedSecondary",
                "outlinedSizeLarge": "MuiButton-outlinedSizeLarge",
                "outlinedSizeSmall": "MuiButton-outlinedSizeSmall",
                "root": "MuiButton-root",
                "sizeLarge": "MuiButton-sizeLarge",
                "sizeSmall": "MuiButton-sizeSmall",
                "startIcon": "MuiButton-startIcon",
                "text": "MuiButton-text",
                "textPrimary": "MuiButton-textPrimary",
                "textSecondary": "MuiButton-textSecondary",
                "textSizeLarge": "MuiButton-textSizeLarge",
                "textSizeSmall": "MuiButton-textSizeSmall",
              }
            }
            onClick={[Function]}
            startIcon={<Memo(AddIcon) />}
          >
            <WithStyles(ForwardRef(ButtonBase))
              className="MuiButton-root MuiButton-text Component-root-1"
              component="button"
              disabled={false}
              focusRipple={true}
              focusVisibleClassName="Mui-focusVisible"
              onClick={[Function]}
              type="button"
            >
              <ForwardRef(ButtonBase)
                className="MuiButton-root MuiButton-text Component-root-1"
                classes={
                  Object {
                    "disabled": "Mui-disabled",
                    "focusVisible": "Mui-focusVisible",
                    "root": "MuiButtonBase-root",
                  }
                }
                component="button"
                disabled={false}
                focusRipple={true}
                focusVisibleClassName="Mui-focusVisible"
                onClick={[Function]}
                type="button"
              >
                <button
                  className="MuiButtonBase-root MuiButton-root MuiButton-text Component-root-1"
                  disabled={false}
                  onBlur={[Function]}
                  onClick={[Function]}
                  onDragLeave={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  onMouseDown={[Function]}
                  onMouseLeave={[Function]}
                  onMouseUp={[Function]}
                  onTouchEnd={[Function]}
                  onTouchMove={[Function]}
                  onTouchStart={[Function]}
                  tabIndex={0}
                  type="button"
                >
                  <span
                    className="MuiButton-label"
                  >
                    <span
                      className="MuiButton-startIcon MuiButton-iconSizeMedium"
                    >
                      <ForwardRef>
                        <WithStyles(ForwardRef(SvgIcon))>
                          <ForwardRef(SvgIcon)
                            classes={
                              Object {
                                "colorAction": "MuiSvgIcon-colorAction",
                                "colorDisabled": "MuiSvgIcon-colorDisabled",
                                "colorError": "MuiSvgIcon-colorError",
                                "colorPrimary": "MuiSvgIcon-colorPrimary",
                                "colorSecondary": "MuiSvgIcon-colorSecondary",
                                "fontSizeInherit": "MuiSvgIcon-fontSizeInherit",
                                "fontSizeLarge": "MuiSvgIcon-fontSizeLarge",
                                "fontSizeSmall": "MuiSvgIcon-fontSizeSmall",
                                "root": "MuiSvgIcon-root",
                              }
                            }
                          >
                            <svg
                              aria-hidden={true}
                              className="MuiSvgIcon-root"
                              focusable="false"
                              viewBox="0 0 24 24"
                            >
                              <path
                                d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                              />
                            </svg>
                          </ForwardRef(SvgIcon)>
                        </WithStyles(ForwardRef(SvgIcon))>
                      </ForwardRef>
                    </span>
                    Add task
                  </span>
                  <WithStyles(memo)
                    center={false}
                  >
                    <ForwardRef(TouchRipple)
                      center={false}
                      classes={
                        Object {
                          "child": "MuiTouchRipple-child",
                          "childLeaving": "MuiTouchRipple-childLeaving",
                          "childPulsate": "MuiTouchRipple-childPulsate",
                          "ripple": "MuiTouchRipple-ripple",
                          "ripplePulsate": "MuiTouchRipple-ripplePulsate",
                          "rippleVisible": "MuiTouchRipple-rippleVisible",
                          "root": "MuiTouchRipple-root",
                        }
                      }
                    >
                      <span
                        className="MuiTouchRipple-root"
                      >
                        <TransitionGroup
                          childFactory={[Function]}
                          component={null}
                          exit={true}
                        />
                      </span>
                    </ForwardRef(TouchRipple)>
                  </WithStyles(memo)>
                </button>
              </ForwardRef(ButtonBase)>
            </WithStyles(ForwardRef(ButtonBase))>
          </ForwardRef(Button)>
        </WithStyles(ForwardRef(Button))>
      </Component>
    </Styled(Component)>
  </TasksAddButton>
</Provider>
`;
